### ISSUE TITLE
Unable to compile source code into an executable

### ISSUE THREAD

**Issue by sim-brar** [martinlindhe/wmi_exporter#116]
Hi there, 

I recently created a custom collector on wmi_exporter using your collector-generator tool. While this wasn't too difficult to figure out, I am unable to compile the source files and get the below error message:

`C:\Users\username\go\src\wmi_exporter-master\exporter.go:85: cannot use ch (type chan<- "wmi_exporter-master/vendor/github.com/prometheus/client_golang/prometheus".Metric) as type chan<- "github.com/martinlindhe/wmi_exporter/vendor/github.com/prometheus/client_golang/prometheus".Metric in argument to c.Collect`

Would you be able to assist me?

Thanks!

**Comment by martinlindhe** [comment]
Hi @trap167 and thanks for reaching out

It seems we have forgotten to update the collector-generator properly after some recent api changes in prometheus. Thanks for letting us know, looking into fixing it

**Comment by trap167** [comment]
Hi Martin,

Absolutely :)

WMI class: Win32_PerfFormattedData_PerfDisk_LogicalDisk

Collector code:

```
// returns data points from Win32_PerfFormattedData_PerfDisk_LogicalDisk
// <add link to documentation here> - Win32_PerfFormattedData_PerfDisk_LogicalDisk class
package collector

import (
	"log"

	"github.com/StackExchange/wmi"
	"github.com/prometheus/client_golang/prometheus"
)

func init() {
	Factories["disk_logicaldisk"] = NewDisk_LogicalDiskCollector
}

// A Disk_LogicalDiskCollector is a Prometheus collector for WMI Win32_PerfFormattedData_PerfDisk_LogicalDisk metrics
type Disk_LogicalDiskCollector struct {
	AvgDiskBytesPerRead     *prometheus.Desc
	AvgDiskBytesPerTransfer *prometheus.Desc
	AvgDiskBytesPerWrite    *prometheus.Desc
	AvgDiskQueueLength      *prometheus.Desc
	AvgDiskReadQueueLength  *prometheus.Desc
	AvgDisksecPerRead       *prometheus.Desc
	AvgDisksecPerTransfer   *prometheus.Desc
	AvgDisksecPerWrite      *prometheus.Desc
	AvgDiskWriteQueueLength *prometheus.Desc
	CurrentDiskQueueLength  *prometheus.Desc
	DiskBytesPersec         *prometheus.Desc
	DiskReadBytesPersec     *prometheus.Desc
	DiskReadsPersec         *prometheus.Desc
	DiskTransfersPersec     *prometheus.Desc
	DiskWriteBytesPersec    *prometheus.Desc
	DiskWritesPersec        *prometheus.Desc
	FreeMegabytes           *prometheus.Desc
	PercentDiskReadTime     *prometheus.Desc
	PercentDiskTime         *prometheus.Desc
	PercentDiskWriteTime    *prometheus.Desc
	PercentFreeSpace        *prometheus.Desc
	PercentIdleTime         *prometheus.Desc
	SplitIOPerSec           *prometheus.Desc
}

// NewDisk_LogicalDiskCollector ...
func NewDisk_LogicalDiskCollector() (Collector, error) {
	const subsystem = "disk_logicaldisk"
	return &Disk_LogicalDiskCollector{
		AvgDiskBytesPerRead: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "avg_disk_bytes_per_read"),
			"(AvgDiskBytesPerRead)",
			nil,
			nil,
		),
		AvgDiskBytesPerTransfer: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "avg_disk_bytes_per_transfer"),
			"(AvgDiskBytesPerTransfer)",
			nil,
			nil,
		),
		AvgDiskBytesPerWrite: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "avg_disk_bytes_per_write"),
			"(AvgDiskBytesPerWrite)",
			nil,
			nil,
		),
		AvgDiskQueueLength: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "avg_disk_queue_length"),
			"(AvgDiskQueueLength)",
			nil,
			nil,
		),
		AvgDiskReadQueueLength: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "avg_disk_read_queue_length"),
			"(AvgDiskReadQueueLength)",
			nil,
			nil,
		),
		AvgDisksecPerRead: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "avg_disksec_per_read"),
			"(AvgDisksecPerRead)",
			nil,
			nil,
		),
		AvgDisksecPerTransfer: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "avg_disksec_per_transfer"),
			"(AvgDisksecPerTransfer)",
			nil,
			nil,
		),
		AvgDisksecPerWrite: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "avg_disksec_per_write"),
			"(AvgDisksecPerWrite)",
			nil,
			nil,
		),
		AvgDiskWriteQueueLength: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "avg_disk_write_queue_length"),
			"(AvgDiskWriteQueueLength)",
			nil,
			nil,
		),
		CurrentDiskQueueLength: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "current_disk_queue_length"),
			"(CurrentDiskQueueLength)",
			nil,
			nil,
		),
		DiskBytesPersec: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "disk_bytes_persec"),
			"(DiskBytesPersec)",
			nil,
			nil,
		),
		DiskReadBytesPersec: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "disk_read_bytes_persec"),
			"(DiskReadBytesPersec)",
			nil,
			nil,
		),
		DiskReadsPersec: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "disk_reads_persec"),
			"(DiskReadsPersec)",
			nil,
			nil,
		),
		DiskTransfersPersec: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "disk_transfers_persec"),
			"(DiskTransfersPersec)",
			nil,
			nil,
		),
		DiskWriteBytesPersec: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "disk_write_bytes_persec"),
			"(DiskWriteBytesPersec)",
			nil,
			nil,
		),
		DiskWritesPersec: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "disk_writes_persec"),
			"(DiskWritesPersec)",
			nil,
			nil,
		),
		FreeMegabytes: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "free_megabytes"),
			"(FreeMegabytes)",
			nil,
			nil,
		),
		PercentDiskReadTime: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "percent_disk_read_time"),
			"(PercentDiskReadTime)",
			nil,
			nil,
		),
		PercentDiskTime: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "percent_disk_time"),
			"(PercentDiskTime)",
			nil,
			nil,
		),
		PercentDiskWriteTime: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "percent_disk_write_time"),
			"(PercentDiskWriteTime)",
			nil,
			nil,
		),
		PercentFreeSpace: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "percent_free_space"),
			"(PercentFreeSpace)",
			nil,
			nil,
		),
		PercentIdleTime: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "percent_idle_time"),
			"(PercentIdleTime)",
			nil,
			nil,
		),
		SplitIOPerSec: prometheus.NewDesc(
			prometheus.BuildFQName(Namespace, subsystem, "split_io_per_sec"),
			"(SplitIOPerSec)",
			nil,
			nil,
		),
	}, nil
}

// Collect sends the metric values for each metric
// to the provided prometheus Metric channel.
func (c *Disk_LogicalDiskCollector) Collect(ch chan<- prometheus.Metric) error {
	if desc, err := c.collect(ch); err != nil {
		log.Println("[ERROR] failed collecting disk_logicaldisk metrics:", desc, err)
		return err
	}
	return nil
}

type Win32_PerfFormattedData_PerfDisk_LogicalDisk struct {
	Name string

	AvgDiskBytesPerRead     uint64
	AvgDiskBytesPerTransfer uint64
	AvgDiskBytesPerWrite    uint64
	AvgDiskQueueLength      uint64
	AvgDiskReadQueueLength  uint64
	AvgDisksecPerRead       uint32
	AvgDisksecPerTransfer   uint32
	AvgDisksecPerWrite      uint32
	AvgDiskWriteQueueLength uint64
	CurrentDiskQueueLength  uint32
	DiskBytesPersec         uint64
	DiskReadBytesPersec     uint64
	DiskReadsPersec         uint32
	DiskTransfersPersec     uint32
	DiskWriteBytesPersec    uint64
	DiskWritesPersec        uint32
	FreeMegabytes           uint32
	PercentDiskReadTime     uint64
	PercentDiskTime         uint64
	PercentDiskWriteTime    uint64
	PercentFreeSpace        uint32
	PercentIdleTime         uint64
	SplitIOPerSec           uint32
}

func (c *Disk_LogicalDiskCollector) collect(ch chan<- prometheus.Metric) (*prometheus.Desc, error) {
	var dst []Win32_PerfFormattedData_PerfDisk_LogicalDisk
	q := wmi.CreateQuery(&dst, "")
	if err := wmi.Query(q, &dst); err != nil {
		return nil, err
	}

	ch <- prometheus.MustNewConstMetric(
		c.AvgDiskBytesPerRead,
		prometheus.GaugeValue,
		float64(dst[0].AvgDiskBytesPerRead),
	)

	ch <- prometheus.MustNewConstMetric(
		c.AvgDiskBytesPerTransfer,
		prometheus.GaugeValue,
		float64(dst[0].AvgDiskBytesPerTransfer),
	)

	ch <- prometheus.MustNewConstMetric(
		c.AvgDiskBytesPerWrite,
		prometheus.GaugeValue,
		float64(dst[0].AvgDiskBytesPerWrite),
	)

	ch <- prometheus.MustNewConstMetric(
		c.AvgDiskQueueLength,
		prometheus.GaugeValue,
		float64(dst[0].AvgDiskQueueLength),
	)

	ch <- prometheus.MustNewConstMetric(
		c.AvgDiskReadQueueLength,
		prometheus.GaugeValue,
		float64(dst[0].AvgDiskReadQueueLength),
	)

	ch <- prometheus.MustNewConstMetric(
		c.AvgDisksecPerRead,
		prometheus.GaugeValue,
		float64(dst[0].AvgDisksecPerRead),
	)

	ch <- prometheus.MustNewConstMetric(
		c.AvgDisksecPerTransfer,
		prometheus.GaugeValue,
		float64(dst[0].AvgDisksecPerTransfer),
	)

	ch <- prometheus.MustNewConstMetric(
		c.AvgDisksecPerWrite,
		prometheus.GaugeValue,
		float64(dst[0].AvgDisksecPerWrite),
	)

	ch <- prometheus.MustNewConstMetric(
		c.AvgDiskWriteQueueLength,
		prometheus.GaugeValue,
		float64(dst[0].AvgDiskWriteQueueLength),
	)

	ch <- prometheus.MustNewConstMetric(
		c.CurrentDiskQueueLength,
		prometheus.GaugeValue,
		float64(dst[0].CurrentDiskQueueLength),
	)

	ch <- prometheus.MustNewConstMetric(
		c.DiskBytesPersec,
		prometheus.GaugeValue,
		float64(dst[0].DiskBytesPersec),
	)

	ch <- prometheus.MustNewConstMetric(
		c.DiskReadBytesPersec,
		prometheus.GaugeValue,
		float64(dst[0].DiskReadBytesPersec),
	)

	ch <- prometheus.MustNewConstMetric(
		c.DiskReadsPersec,
		prometheus.GaugeValue,
		float64(dst[0].DiskReadsPersec),
	)

	ch <- prometheus.MustNewConstMetric(
		c.DiskTransfersPersec,
		prometheus.GaugeValue,
		float64(dst[0].DiskTransfersPersec),
	)

	ch <- prometheus.MustNewConstMetric(
		c.DiskWriteBytesPersec,
		prometheus.GaugeValue,
		float64(dst[0].DiskWriteBytesPersec),
	)

	ch <- prometheus.MustNewConstMetric(
		c.DiskWritesPersec,
		prometheus.GaugeValue,
		float64(dst[0].DiskWritesPersec),
	)

	ch <- prometheus.MustNewConstMetric(
		c.FreeMegabytes,
		prometheus.GaugeValue,
		float64(dst[0].FreeMegabytes),
	)

	ch <- prometheus.MustNewConstMetric(
		c.PercentDiskReadTime,
		prometheus.GaugeValue,
		float64(dst[0].PercentDiskReadTime),
	)

	ch <- prometheus.MustNewConstMetric(
		c.PercentDiskTime,
		prometheus.GaugeValue,
		float64(dst[0].PercentDiskTime),
	)

	ch <- prometheus.MustNewConstMetric(
		c.PercentDiskWriteTime,
		prometheus.GaugeValue,
		float64(dst[0].PercentDiskWriteTime),
	)

	ch <- prometheus.MustNewConstMetric(
		c.PercentFreeSpace,
		prometheus.GaugeValue,
		float64(dst[0].PercentFreeSpace),
	)

	ch <- prometheus.MustNewConstMetric(
		c.PercentIdleTime,
		prometheus.GaugeValue,
		float64(dst[0].PercentIdleTime),
	)

	ch <- prometheus.MustNewConstMetric(
		c.SplitIOPerSec,
		prometheus.GaugeValue,
		float64(dst[0].SplitIOPerSec),
	)

	return nil, nil
}
```

**Comment by martinlindhe** [comment]
Hello @trap167 

The code you posted compiles fine here, using the git master branch of wmi_exporter.
I cannot reproduce your issue.

**Comment by carlpett** [comment]
I agree with @martinlindhe, something is off with the vendoring. What version of go are you running?

As an additional question, is there something you are missing in the existing `logical_disk` collector since you are generating a new one? Also, `Win32_PerfFormattedData_PerfDisk_LogicalDisk` is a "formatted" class, which means that the data is not the raw collected data, but has been massaged a bit by Windows to fit the needs of Perfmon. This is _probably_ not what you want?

**Comment by trap167** [comment]
Hi @martinlindhe and @carlpett,

I followed your advice and tried compiling with a clean checkout of wmi_exporter. This worked perfectly and I was able to begin collecting Win32_PerfFormattedData_PerfDisk_LogicalDisk metrics.

Here are my answers to your queries:

1) I am running go 1.9
2) The existing logical_disk collector works fine for most of our servers, except for those running Windows Server 2008 (reports incorrect disk capacity information). I did some research and found that the Win32_PerfFormattedData_PerfDisk_LogicalDisk collector works fine as a substitute.

And a question for you both:

The wmi_exporter.exe that I compiled with the custom collector does not run as a service. I am wondering if this means I compiled incorrectly, or if only your .msi releases are able to run as services?
